BB-1744: Bump pymongo version and fix course export issue

giovannicimolin opened the pull request:
This PR upgrades the Pymongo driver used by `edx-platform` to a newer version that supports newer MongoDB versions.

**JIRA tickets**: [OSPR-3757](https://openedx.atlassian.net/browse/OSPR-3757)

**Discussions**: 
* [ARCH-612](https://openedx.atlassian.net/browse/ARCH-612)
* [Discuss forum post](https://discuss.openedx.org/t/upgrade-to-mongo-3-6-any-time-mongodb-atlas-is-no-longer-supporting-mongo-3-4/)

**Sandbox URLs**: 
Users: usual testing users (staff, honor, verified, etc)

* MongoDB 3.2 (using this branch)
LMS: https://pr21197.sandbox.opencraft.hosting/
Studio: https://studio.pr21197.sandbox.opencraft.hosting/

* MongoDB 3.6 (backported to Ironwood)
LMS: https://pymongo-mongo36-test.opencraft.hosting/
Studio: https://pymongo-mongo36-test.opencraft.hosting/

* MongoDB 4.0 (backported changes to Ironwood)
LMS: https://pymongo-mongo40-test.opencraft.hosting/
Studio: https://studio.pymongo-mongo40-test.opencraft.hosting/

**Merge deadline**: Before the end of the year.

**Testing instructions**:
1. Perform a full testing checklist on the sandbox instances.

**Author notes and concerns**:

1. The LMS and Studio are working as expected, but I didn't run any performance benchmarks on the newer driver against the old one.

**Reviewers**
- [x] @viadanna or @lgp171188 
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @giovannicimolin! I've created [OSPR-3757](https://openedx.atlassian.net/browse/OSPR-3757) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

natabene commented:
@giovannicimolin Thank you for your contribution. Please let me know once it is ready for our review.

feanil commented:
@giovannicimolin it looks like most of the test failures are in test setup because the way we used to drop databases no longer works.  Do you have enough context to fix this?

giovannicimolin commented:
@feanil I have some context, but limited budget to do this. I can try to work out as many issues as I can.

viadanna commented on common/lib/xmodule/xmodule/modulestore/mongo/draft.py:
```diff
@@ -651,7 +651,7 @@ def _internal(tier):
             bulk_record = self._get_bulk_ops_record(root_usages[0].course_key)
             bulk_record.dirty = True
-            self.collection.remove({'_id': {'$in': to_be_deleted}}, safe=self.collection.safe)
+            self.collection.remove({'_id': {'$in': to_be_deleted}})
```
Nit: `collection.remove` has been deprecated, can you change this to `delete_many`?

feanil commented:
@giovannicimolin congrats on getting all the tests passing! Nice job!

viadanna commented on requirements/edx/base.txt:
```diff
@@ -192,7 +192,7 @@ pygments==2.4.2
 pyjwkest==1.3.2
 pyjwt==1.5.2
-pymongo==2.9.1
+pymongo==3.8.0
```
`pymongo==3.9.0` has been released, do you think it would be hard to upgrade?
https://api.mongodb.com/python/current/changelog.html#changes-in-version-3-9-0

giovannicimolin replied:
Done. Note that `3.9.0` might not have an available version when backporting to older releases.

giovannicimolin commented:
@feanil Thanks! I think this is almost ready for review. We're doing some tests on the sandbox to ensure everything is working.
Also, thanks for the heads up about the CI.

viadanna commented on common/lib/xmodule/xmodule/modulestore/mongo/base.py:
```diff
@@ -627,14 +620,14 @@ def _drop_database(self, database=True, collections=True, connections=True):
             self.collection.drop()
         else:
-            self.collection.remove({})
+            self.collection.delete_many({})
```
Thanks! I sifted the branch and found no other usages.

giovannicimolin commented:
@viadanna Thanks for the fast review!

@natabene @feanil This is ready for edX's review.

natabene commented:
@giovannicimolin @pomegranited Can you please prioritize it in our shared Google doc?

giovannicimolin commented:
@natabene Sure, done!

giovannicimolin commented:
jenkins run all

giovannicimolin commented:
jenkins run py35-django111/python

giovannicimolin commented:
@feanil I'm having issues figuring out why the tests are failing for the Python 3 unit tests specifically for `lms/djangoapps/verify_student`. It seems like the pymongo driver is failing to connect to mongodb server to create a new course using `CourseFactory`. 
The same tests work for Python 2, so I'm thinking this might be an issue in the exception handler [here](https://github.com/edx/MongoDBProxy/blob/master/mongodb_proxy.py#L40) that isn't working on Python 3.
Can you take some time to take a look at this?

giovannicimolin commented:
Is there a way of running the Python 3 test suite locally?

feanil commented:
@giovannicimolin in devstack you can run it via tox

`tox -e py35-django111 -- pytest /path/to/test.py`

You can also just re-run the python3.5 tests on github via commands like
`jenkins run py35-django111 <context>` eg. `jenkins run py35-django111 python`

giovannicimolin commented:
@feanil Thanks a lot for the tip! It was really useful. I will try to debug the issue and find a solution.

giovannicimolin commented:
I boiled this down to class inheritance issues caused by differences on python 2 and 3. 
On python 2 the both `MockS3Mixin` and `ModuleStoreTestCase` `.setUp()` methods are called, while on python 3, only the one from `MockS3Mixin` is called.
I'm not exactly sure how to fix it yet, but I'll try again later this week.

giovannicimolin commented:
@feanil I've fixed some test by inverting the inheritance order, but I wasn't able to fix the one from the management command. Since this is happening because some inheritance issues on the `ModuleStoreTestCase` class, and are not affecting the actual behavior, I added the `py2_only` decorator to the failing test. 
We don't have a lot of time left on the epic to work on this, but I hope this is enough to bump the driver version to one that runs on a supported mongodb version.

feanil commented on lms/djangoapps/verify_student/management/commands/tests/test_send_verification_expiry_email.py:
```diff
@@ -206,11 +206,14 @@ def test_not_enrolled_in_verified_course(self):
-        email again
+        email again.
+
+        Does not work on python 3 with the latest mongo driver version due to class inheritance issues.
```
Can you provide more details on this?  What is the actual error that we see?  Given that this is essentially deleting the test going forward, I want to understand what the issues are.

The `@py2_only` decorator was meant to be used during transition and is not meant to be permanent.

giovannicimolin replied:
@feanil Sorry, for the delay in replying, I missed this ping.
Using the new `pymongo` driver, the `ModuleStoreTestCase` doesn't correctly set up the module store for testing.
On the other set of tests, this was fixed by switching the inheritance order (see https://github.com/edx/edx-platform/pull/21197/files/5983d5dd42ca8f0f6e1347d04ccbcc4a28482a92#diff-16556c13da8f491bf5a44a4a628ccc25R26), but the workaroung didn't work on this one.
When I tried swapping the inheritance order here, the S3 portion of the code started failing, and this led me to believe there is something wrong with the class inheritance on the test cases classes (`ModuleStoreTestCase`, `MockS3Mixin`). 
I've spent some time on the issue and couldn't make the test work. :(

That said, the tests pass on Python 2, and the feature seems to be working fine (from limited testing), so I used the `py2_only` decorator to get the tests green and check if its possible to drop this test.

feanil replied:
Got it,  the MockS3Mixin has a lot of issues and we had a bunch of trouble with it with other tests as well.  Given that we're going to be dropping python 2 support, I think we can delete this test case if it's being flaky.

@giovannicimolin can you please delete this test in a separate PR as a follow up?

giovannicimolin replied:
@feanil I've created https://github.com/edx/edx-platform/pull/22618 to delete the flaky tests.

edx-webhook commented:
@giovannicimolin 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/21197&entry.1484655318=https://github.com/giovannicimolin&entry.752974735=2019-07-23+20:52&entry.1917517419=2019-12-26+16:16&entry.2133058324=Yes) so we can improve your experience in the future.

giovannicimolin commented:
@feanil Thanks for merging the PR! :slightly_smiling_face: 

I've created this follow up PR to delete the flaky test: https://github.com/edx/edx-platform/pull/22618.

giovannicimolin commented:
@feanil Do you think this PR might be causing issues in the production deployment? Is there anything I can do to help?
